File: PythonCodes/Fig09_04.py
#%%
"""
Created on Jan 20 2019
Smoothness of the payoff and Monte Carlo convergence
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

def mainCalculation():
    NoOfPaths = range(100,10000,100)
    
    # Define two payoff functions

    g1 = lambda x: (x > 0.0)
    g2 = lambda x: st.norm.cdf(x)
    
    # Plot the g1 and g2 functions

    x =np.linspace(-5.0,5.0,100)
    plt.figure(1)
    plt.plot(x,g1(x))
    plt.plot(x,g2(x))
    plt.grid()
    plt.xlabel('x')
    plt.ylabel('gi(x)')
    plt.legend(['g1(x)','g2(x)'])

    # Store the errors 

    error1 = np.zeros([len(NoOfPaths),1])
    error2 = np.zeros([len(NoOfPaths),1])
    
    for idx, noOfPaths in enumerate(NoOfPaths):
        W1 = np.random.normal(0.0,1.0,[noOfPaths,1])
        if noOfPaths > 1:
            W1 = (W1 - np.mean(W1)) / np.std(W1)
        approx1 = np.mean(g1(W1))
        approx2 = np.mean(g2(W1))
        error1[idx] = approx1 - 0.5
        error2[idx] = approx2 - 0.5
    
    # Plot convergence results

    plt.figure(2)
    plt.plot(NoOfPaths,error1)
    plt.plot(NoOfPaths,error2)
    plt.grid()
    plt.xlabel('No of paths')
    plt.ylabel('error')
    plt.legend(['error 1','error 2'])

mainCalculation()

File: MatlabCodes/Chapter 09/Fig09_04.m
function PayoffConvergence
close all;clc
randn('seed',3) %6

% Define two payoff functions

g1 = @(x)1*(x>0);
g2 = @(x)normcdf(x);

% Maturity time

T = 1;

% Range of paths

NoOfPathsV=100:100:10000;
i=1;

% Storage of the error

error1 = zeros([length(NoOfPathsV),1]);
error2 = zeros([length(NoOfPathsV),1]);

idx = 1;
for NoOfPaths=NoOfPathsV
    Z  =random('normal',0,1,[NoOfPaths,1]);
    Z  =(Z-mean(Z))/std(Z);
    W1 =sqrt(T).*Z;
    
    % Expectation

    val1 = mean(g1(W1));
    val2 = mean(g2(W1));
    
    % For both cases the exact solution is equal to 0.5

    error1(idx)=val1 - 0.5;
    error2(idx)=val2 - 0.5;
    idx=idx+1;
end

figure(1)
x=-5:0.1:5;
plot(x,g1(x),'.--','linewidth',1.5);
hold on
plot(x,g2(x),'-r','linewidth',1.5);
grid on;
xlabel('x');
ylabel('g_i(x)');
leg=legend('$$g_1(x)$$','$$g_2(x)$$');
set(leg,'Interpreter','latex');
tit=title('Payoffs $$g_1(x)$$ and $$g_2(x)$$');
set(tit,'Interpreter','latex');
%axis([-5.5,5.5,-0.05,1.05]);

figure(2)
plot(NoOfPathsV,error1,'.--','linewidth',2);
hold on
plot(NoOfPathsV,error2,'-r','linewidth',2);
grid on;
xlabel('N');
ylabel('error $$\tilde{c}_N$$','Interpreter','latex');
leg=legend('error $$\tilde{c}_N^1$$','error $$\tilde{c}_N^2$$');
set(leg,'Interpreter','latex');
tit=title('errors $$\tilde{c}_N^1$$ and $$\tilde{c}_N^2$$');
set(tit,'Interpreter','latex');
